### PR TITLE
Enhanced Goals

### LINKED ISSUE #27: Improve Goals with new functionality
Edit goals so that goals that address objects not yet created can be specified (e.g. goal for fill water should be in(water, pot) when 'water' is not in env file)

### PR DESCRIPTION
**Pull request by lsuyean**
## Overview

Previously, goals needed to be specified using objects declared within the problem file. This means that  goals that involved created objects were impossible (like checking if a pot has been filled up with water, a created object). In this PR, I enhance goals to be able to include goals that could achieve this. 

When specifying goals that involve created objects in the problem file, we can only specify them as part of a combination predicate (since there is no unique id for an object that does not yet exist). We do so as follows:

```
{
    "predicate": "in",
    "args": ["water", "pot"],
    "ids": [1, 2] 
}
```

## Changes Made

- Edited state.py so that goals do not need to include objects defined in domain
- Edited builder.py and env.py to create objects for goals with objects not defined in domain
- Updated test files with new enhanced goals
- Fixed bug that wrongly created predicates for station location
- Fixed bug that allowed pot to be filled when not in sink (added a conditional sfx)
- Fixed bug that does not allow for proper creation of multiple objects (added new test case base_fill_two_pots) 

## Test Coverage

Tested manually

## Related PRs or Issues (delete if not applicable)

Closes #27

### CODE REVIEW

File: environments/env_generator/builder.py [hunk 1]
```diff
@@ -300,13 +303,14 @@ def create_unique_and_combination_preds(environment_dict):
                     # Get all entities to prepare the combination
                     arg_entities = list(filter(lambda entity: arg in entity["name"], environment_dict[entity_field]))
                     arg_entity_names = list(map(lambda entity: entity["name"], arg_entities))
-                    combination_dict[arg]['entities'] = arg_entity_names
+                    combination_dict[arg]['entities'] = arg_entity_names if arg_entity_names else [arg + "1"]
```
Comment (chalo2000) [inline environments/env_generator/builder.py:306]: This will likely limit us in the future with creating multiple objects that are part of a goal

**Review by chalo2000 — CHANGES_REQUESTED** [review]
Leaving some temporary feedback in response to our conversation

File: backend/state.py [hunk 2]
```diff
@@ -161,8 +161,8 @@ def initialize(self, domain, objects, true_predicates, all_goals, special_effect
         # check if goal predicates are defined in domain
         for goal_set in all_goals:
             for goal in goal_set:
-                if goal not in predicates:
-                    raise ValueError(f"Predicate {goal} is not defined in the domain.")
+                if goal.name not in list(map(lambda x: x.name, domain.predicates)):
```
[the remark this answers is not in the document]
Comment (lsuyean) [reply backend/state.py:164]: defined list outside of loop

File: backend/state.py [hunk 3]
```diff
@@ -198,13 +198,9 @@ def get_predicate_value(self, predicate):
...
         Returns:
-            bool: True if the predicate is True in the state, False otherwise.
-
-        Raises:
-            AssertionError: If the predicate is not in the state.
+            bool: True if the predicate is True in the state, False otherwise. 
         """
-        assert predicate in self.predicates
-        return self.predicates[predicate]
+        return self.predicates[predicate] if predicate in self.predicates else False
```
[the remark this answers is not in the document]
Comment (lsuyean) [reply backend/state.py:203]: Documentation updated

File: environments/env_generator/builder.py [hunk 4]
```diff
@@ -181,13 +182,14 @@ def build_player_location_predicates(environment_dict):
...
     for player in environment_dict["players"]:
         match = False
-        for item in environment_dict["items"]:
-            if player["x"] == item["x"] and player["y"] == item["y"]:
-                predicates_str += f"    (has {player['name']} {item['name']})\n"
-                match = True
-                break
-        if not match:
-            predicates_str += f"    (nothing {player['name']})\n"
+        if environment_dict.get("items"):
```
[the remark this answers is not in the document]
Comment (lsuyean) [reply environments/env_generator/builder.py:185]: updated accordingly

File: environments/env_generator/builder.py [hunk 5]
```diff
@@ -228,12 +230,13 @@ def build_stacking_predicates(environment_dict):
     stacks = {}
     # Sort items into stacks ordered by stacking order
     sorting_key = lambda item: item["stack-level"]
-    for item in environment_dict["items"]:
-        for station in environment_dict["stations"]:
-            if item["x"] == station["x"] and item["y"] == station["y"]:
-                stacks[station["name"]] = stacks.get(station["name"], []) + [item]
-                stacks[station["name"]].sort(key=sorting_key)
-                break
+    if environment_dict.get("items"):
```
[the remark this answers is not in the document]
Comment (lsuyean) [reply environments/env_generator/builder.py:233]: Updated

File: robotouille/env.py [hunk 6]
```diff
@@ -84,11 +84,11 @@ def build_station_location_predicates(environment_dict):
     predicates = []
     for station in environment_dict["stations"]:
         station_obj = Object(station["name"], "station")
-        for field in ["items", "players", "containers"]:
+        match = False
+        for field in ["players", "items", "containers"]:
```
[the remark this answers is not in the document]
Comment (lsuyean) [reply robotouille/env.py:88]: Separated players loop from items and containers loop for clarity

File: robotouille/env.py [hunk 7]
```diff
@@ -98,9 +98,13 @@ def build_station_location_predicates(environment_dict):
                     pred = Predicate().initialize(predicate, [field[:-1], "station"], [obj, station_obj])
                     predicates.append(pred)
                     match = True
-            if not match:
+            if not match and field == "players":
                 pred = Predicate().initialize(no_match_predicate, ["station"], [station_obj])
                 predicates.append(pred)
+            match = False if field == "players" else match
```
[the remark this answers is not in the document]
Comment (lsuyean) [reply robotouille/env.py:104]: Separated players loop from items and containers loop

**Pull request merged**